### DISCUSSION TITLE
Accept request body of "application/octet-stream" type?

### DISCUSSION THREAD

**Discussion by hongquan** [litestar-org/litestar#3038] [Q&A] [answered]
Hi.

Is there any way to define handler that accept request body of "application/octet-stream" type?

I tried

```py
async def flash_firmware(self, data: Annotated[UploadFile, Body(media_type='application/octet-stream')]) -> FirmwareUploadResponse
```
and
```py
async def flash_firmware(self, data: Annotated[bytes, Body(media_type='application/octet-stream')]) -> FirmwareUploadResponse
```

But always get `JSON is malformed: invalid character (byte 0)` error reponse. Litestar keeps trying to decode request body as JSON.

The reason I want this feature is to implement OTA firmware update for ESP32 chip. Due to limit of RAM, I cannot afford any deserializtion, nor validation feature, and can only expect the client to upload proper binary file. The Python code here is to make a mock server for developing web frontend. The actual backend code will be in Rust, based on ESP-IDF framework.

**Comment by provinzkraut** [comment]
Can you post a full example of your code with the first example? That should work afaik.

Anyway, if you don't want *all* validation, you'll have to skip injecting the body via the `body` parameter, since this will always validate, even if it's just `bytes`. You can get the raw body though [`Request.body()`](https://docs.litestar.dev/latest/reference/connection.html#litestar.connection.Request.body):

```python
async def flash_firmware(request: Request) -> FirmwareUploadResponse:
    body = await request.body()
```

**Reply by Alc-Alc** [reply, marked as answer]
to add to this, you can directly have `body` injected as mentioned here https://docs.litestar.dev/latest/usage/routing/handlers.html#reserved-keyword-arguments

```python
async def flash_firmware(body: bytes) -> FirmwareUploadResponse:
    body
```

**Reply by provinzkraut** [reply]
Yeah.. Do that instead. Thanks @Alc-Alc! I sometimes forget about features we have :)

**Reply by hongquan** [reply]
Thanks, the signature

```py
async def flash_firmware(body: bytes) -> FirmwareUploadResponse:
    body
```
works!
